### ISSUE TITLE
Product page - tech task - Improve Category tree widget to handle huge number of categories

### ISSUE THREAD

**Issue by matks** [PrestaShop/PrestaShop#19281]
On 1st tab you can link a product to multiple categories and there is a category tree to choose the target categories. It probably breaks when there are too many categories.

Goal: check current categories widget and see how it can scale when given thousand of categories input data.

**Comment by zuk3975** [comment]
I've generated categories, assigning them to random parents. Here is what I found (numbers of categories - behavior):
 - < 400 - loads and works fine.
- **500** - reaches php memory limit if its less than 128mb. (raised the memory limit to _512mb_ to proceed). By the way the list is already huge, and from this number it would make sense to avoid expanding every category tree.
- 1500 - loads up fine, but you can already see couple seconds delay until component is usable
- 2000 - recognizable slowing in page load. but still usable.
- 3000 - page load over 10s until component is usable
- **3500** - php memory_limit reached with it being 512mb (raised the memory limit to _1G_ to proceed)
- 4000 - page load (until component is usable) reaches 30s
- 5000 - visible delay of ~1s when expanding/collapsing category list
- **5500-6000** - component starts lagging, timeouts from time to time on page refresh. (by this i assume its unusable).

It still depends alot from machine. Tested on 20Gb ram ubuntu19/apache2/php7.3

**Comment by sakonn** [comment]
I am also facing the issue, product page is completely unusable for shop with more than 5000 categories. It's weird that such a large platform can't handle larger number for categories. 
My quick solution was to disable category widget on the product page to at least somehow enable product editing and then I have created external app which can edit categories through webservice.

**Comment by matks** [comment]
> I am also facing the issue, product page is completely unusable for shop with more than 5000 categories. It's weird that such a large platform can't handle larger number for categories.

This is because, a long time ago, PrestaShop was built for (very) small merchants.

Luckily the software grew a lot and its audience grew to small merchants then to middle merchants so the scalability issues start popping 😄  but this is good news: it means PrestaShop users business size is growing and growing.

An other possible explanation is that, when merchants reach a certain size, they start going into PrestaShop Back office 🤔  . Usually they have an ERP and synchronized their data between their eCommerce website (powered by PrestaShop) and their ERP. So they dont see PrestaShop limitations in the BO because they never go there 😄

**Comment by hibatallahAouadni** [comment]
@matks could you please check this issue #13698
I think it's similar to yours :wink:

**Comment by matks** [comment]
@hibatallahAouadni Yes it's probably the same root issue. Can we keep this one open and close https://github.com/PrestaShop/PrestaShop/issues/13698 as duplicate? Because https://github.com/PrestaShop/PrestaShop/issues/13698 is about the consequences while this ticket is about the cause.

**Comment by hibatallahAouadni** [comment]
@matks Ok I will do it :wink: But could you tell me what's the status of this issue? is it **`To Do`**?

**Comment by soolyyy** [comment]
Is there any "hot-fix" for this ? Or any date when it will be fixed? We came across this problem too and need to solve it.

**Comment by roncarino** [comment]
I would love a "hot-fix" too. Unfortunately, we are facing the same problem with a quite big client and we need to find an "urgent" fix. Prestashop 1.7.5.0

**Comment by thebolet** [comment]
Hi @roncarino,
I just have the same problem, what solution did you find for your project ?

**Comment by roncarino** [comment]
> Hi @roncarino, I just have the same problem, what solution did you find for your project ?

I didn't find any solution really. They are completely re-working the product page for the new Prestashop 8 so I believe there won't be any hot-fix for previous versions. In my case, I just removed the search bar improving the speed loading a little bit but nothing more. They are fine to use a CMD + F through the browser.

**Comment by viv3s** [comment]
My post was duplicate of this too #28759 any improve?

**Comment by viv3s** [comment]
¿Any improve or workarround fix ?

**Comment by psdes** [comment]
Not fixed in 8.0.0 so no impove at all.

**Comment by zuk3975** [comment]
It was improved a little in product page v2.
Now the whole category tree is loaded in a dedicated modal only when you click "add categories" button, so it is at least not loaded everytime when opening product edit form (so in case you have lots of categories, it shouldn't break the whole edition page).
But still in the tree itself there are no lazy loading, it loads all of them at once.

**Comment by Hlavtox** [comment]
Closing as the old product page is gone now.
